[23657] BottomSheet cards refactoring

MichaelReyes opened the pull request:
# Pull Request (PR) Checklist

Thank you for your contribution! Please confirm that you've checked all the boxes below before submitting your PR. Use `[x]` to check a box, e.g., `[x]`, and make sure there's no space around the brackets.

## PR Context

- **Type**: Refactor
- **Issue Link**: [Redmine Issue](https://redmine.buzzhives.com/issues/23657)
- **Risk Factor**: High

## Changes

_Describe your changes in detail, highlighting the problem it solves or the feature it adds._

**Retained Bottom Sheet Implementation**
- Continued using `FrameLayout` + `BottomSheetBehavior` inside `HomeFragment` to support touch interactions with the map behind.

**Introduced Modular Architecture Components**
- **`TKUICardHost`**  
  Interface contract to host fragments inside the bottom sheet.

- **`TKUICardNavigator`**  
  Per-screen navigator classes (e.g., `SearchNavigator`, `RouteNavigator`, `TripNavigator`) that handle navigation logic per card.

- **`TKUICardViewControllerManager`**  
  Utility class that handles fragment transactions and setup logic dynamically through the `FragmentManager`.

**Refactored to Use `TKUICardBaseFragment`**
Each bottom sheet screen now extends `TKUICardBaseFragment` and defines its own:
- `behaviorState`
- `peekHeightResourceValue`
- `isHideable`

This makes each screen self-contained and allows custom behavior configurations.

## Checklist for Reviewers

### Documentation and Code Quality
- [ ] **KDocs Documentation**: Are all changes, new functionalities, and classes documented with KDocs?
- [ ] **Architectural Patterns**: Is there consistent and proper use of architectural patterns (e.g., MVVM, MVP)?

### Testing and Reliability
- [ ] **Unit Testing**: Are there unit tests for all new functionalities and classes?
- [ ] **Emulator and Real Device Testing**: Has the application been tested on both emulators and real devices to ensure compatibility?

### Error Handling and Logging
- [ ] **Error Handling**: Are errors and exceptions caught and handled gracefully, ensuring the app remains stable?
- [ ] **Logging**: Is there proper logging in place for critical errors and information, aiding in debugging and monitoring?


## Testing Procedure

_If applicable, provide steps or commands for testing your changes. This can help reviewers and testers._

- 
- 
-

## Work-in-Progress (WIP)

_List any remaining work or areas that need additional focus. This section can be updated as the work progresses._

- [ ] 
- [ ] 

_Remember to keep this template updated based on the feedback and evolving project standards._

MichaelReyes committed:
[23657] - [WIP] refactor bottom sheet handling to support dynamic configuration and decoupled navigation

MichaelReyes committed:
Merge branch 'refs/heads/task/23642-home-card-items-refactoring' into task/23657-bottom-sheet-cards-refactoring

MichaelReyes committed:
Rename .java to .kt

MichaelReyes committed:
[23642] - TripPreview navigation initialization changes
- TripResultPagerFragment.java convertion to kotlin
- TKUICardNavigator introduction
- update TKUICardViewControllerManager to return committed fragment on showCard
- TripResultPagerFragment view pager implementation to use view pager 2 and added new adapter TripGroupsPagerAdapter
-

MichaelReyes committed:
[23642] - [TripGo] update HomeFragment to implement TKUICardHost getMap function and update popBackStack to include popBackStackImmediate call
- [TripGo] update HomeSearchNavigator to modify calling of SearchCardFragment to use popBackStackImmediate showCard
- [TripGo] update LocationChooserFragment to extend TKUICardBaseFragment and implement it's methods and replace event bus calls to use navigator class
- [TripGo] update RouteCardFragment to extend TKUICardBaseFragment and implement it's methods
- [TripGo] add RouteNavigator for all fragment transaction/navigation related to routing
- [TripGo] update SearchCardFragment to extend TKUICardBaseFragment and implement it's methods and replace event bus calls to use navigator class
- [TripGo] add SearchNavigator for all fragment transaction/navigation related to search
- [TripGo] update TimetableFragment to extend TKUICardBaseFragment and implement it's methods
- [TripGo] add TimetableNavigator for all fragment transaction/navigation related to timetable
- [TripKitUI] update TKUICardHost to modify popBackstack call and add getMap call
- [TripKitUI] update TKUICardNavigator to add handling for configs and parent fragment
- [TripKitUI] update TKUICardViewControllerManager to add get fragment by tag and by id functions

MichaelReyes committed:
[23657] - [TripGo] update RouteCardFragment and remove usages of event bus and replaced it with the navigator class
- [TripGo] update RouteNavigator to add function to load the route to open TripResultsFragment
- [TripGo] update TripResultsFragment to TKUICardBaseFragment and implement its members
- [TripKitUI] update TKUICardBaseFragment to add close action handling

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/controller/trippreviewcontroller/TKUITripPreviewFragment.kt:
```diff
@@ -175,20 +177,23 @@ class TKUITripPreviewFragment : BaseFragment<FragmentTkuiTripPreviewBinding>() {
+        sharedViewModel.apply {
+            observe(pageIndex) {
+                it?.let {
+                    if (!fromPageListener) {
```
Well written code, can improve readability by utilizing `return@observe` but not necessary.

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/generic/bottomsheet/BottomSheetCardsManager.kt:
```diff
@@ -0,0 +1,91 @@
+
+    }
+
+    fun setState(state: Int) {
```
This could benefit from a safety check similar to the one used in `restore()` to guard against invalid states.

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/generic/bottomsheet/BottomSheetCardsManager.kt:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    fun push() {
+        behavior.let {
```
`let` seems to be unused here.

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/generic/bottomsheet/TKUICardViewControllerManager.kt:
```diff
@@ -0,0 +1,71 @@
+            transaction.addToBackStack(fragmentClass.name)
+        }
+
+        transaction.commitAllowingStateLoss()
```
Can result in hard to debug issues.

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/generic/bottomsheet/TKUICardViewControllerManager.kt:
```diff
@@ -0,0 +1,71 @@
+        currentFragment = fragment
+
+        val transaction = fragmentManager.beginTransaction()
+            .replace(contentFrameId, fragment)
```
Adding tag can be of help.

sg-jsonjuliane commented on TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/generic/card/TKUICardDialog.kt:
```diff
@@ -0,0 +1,71 @@
+                behavior.isHideable = false
+
+                // ✅ Delay fragment commit until container is attached
+                Handler(Looper.getMainLooper()).postDelayed({
```
Instead of adding delays, might be better to use `post {}`.

sg-jsonjuliane reviewed (approved):
Left a few comments, but no changes requested — everything looks good.

MichaelReyes committed:
Merge branch 'refs/heads/develop' into task/23657-bottom-sheet-cards-refactoring

# Conflicts:
#	TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/controller/tripdetailsviewcontroller/TKUITripDetailsViewControllerFragment.kt
#	TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/tripresult/TripResultPagerFragment.kt

MichaelReyes committed:
[23657] - [TripKitUI] fixing merge conflicts

MichaelReyes committed:
Merge branch 'refs/heads/develop' into task/23657-bottom-sheet-cards-refactoring

MichaelReyes committed:
[23657] - [TripKit] Fix OkHttp timeout configurations
- [TripGo] Fix WLs build issues after merging changes from the latest develop

MichaelReyes committed:
[23657] - [TripKit] Fix failing test

MichaelReyes committed:
[23657] - Fix navigation from TripDetails screen that causes the user need to press close twice to go back
- Fix trip alert toggle not working

MichaelReyes committed:
[23657] - Navigation checking fixes by properly settings fragment tags when adding to backstack

MichaelReyes committed:
Merge branch 'refs/heads/develop' into task/23657-bottom-sheet-cards-refactoring

MichaelReyes committed:
[23657] - fix for map pin inconsistent behavior and pin location bottomsheet info crash

MichaelReyes committed:
[23657] - update HomeFragment to rename setShowPoiMarkers function from TripKitMapFragment
- update HomeNavigator to change poi marker state restoration trigger from TripDetailsFragment to TripResultsFragment
- add MapComponent for map related components for injection
- update TripDetailsFragment to remove poi markers state update since it's moved to TripResultsFragment
- update TripGoSDK to add MapComponent
- update TripResultsFragment to initialize and pass mapContributor using TripResultMapContributor to TripResultListFragment
- move TripResultMapContributor from TripKitUI to TripGoSDK and use the new TripResultListMapContributor interface from TripKitUI
- add MapExtensions.kt to handle all extensions and constants for Map related stuffs
- update TripKitMapFragment to add handling of passing markerManager to TripResultListMapContributor, handling for showing/hiding markers and getting/creating marker collections
- update TripResultListFragment and update how mapContributor is set and handled
- remove old TripResultListMapContributor.kt that's replaced with TripResultListMapContributor interface
- add TripResultListMapContributor interface so map contributor for TripResultListFragment can be more configurable
- update TripResultMapContributor and allow setting of MarkerManager and only create a new manager if it's null. Also update handling of marker collections

MichaelReyes committed:
Merge branch 'refs/heads/develop' into task/23657-bottom-sheet-cards-refactoring

# Conflicts:
#	tripkit-android

MichaelReyes committed:
[23657] - add timeout to get location call to ensure it won't block user waiting for location data.

MichaelReyes committed:
Merge branch 'refs/heads/develop' into task/23657-bottom-sheet-cards-refactoring

MichaelReyes committed:
[23657] - update tripkit-android and tripkit-android-ui submodule references

MichaelReyes committed:
[24457][24455] - Add dialog to ask to enable location and settings or fallback to proceed without checking location

MichaelReyes committed:
[23657] - convert ScheduledStopRepository to use Room database
- add Room entities: ScheduledStopEntity, LocationEntity, ScheduledStopDownloadHistoryEntity
 - add ScheduledStopWithLocation data class for joined queries
- add ScheduledStopDao with CRUD operations and bounds queries
- add ScheduledStopMapper to convert between Room entities and domain models
- add ScheduledStopDatabase as main Room database class
- update StopsPersistor to create Room entities directly instead of ContentValues
- update ScheduledStopRepository to use Room queries instead of ContentProvider
- add simplified bounds queries without download history requirement
- update Dagger module to provide Room components
- remove cursor-based approach as requested by user

MichaelReyes committed:
[23657] - Fixed ZoomLevel.fromLevel() method that was returning null for zoom levels 8.1f-12.9f
- Updated zoom level ranges: REGION (8.1f-12.9f), HYBRID (13.0f-15.1f), LOCAL (≥15.2f)
- Implemented hybrid approach where 13.0f-15.1f loads both region and local level stops
- Ensured ViewPort.CloseEnough is created for all valid zoom levels to trigger API calls
- Fixed issue where REGION level markers weren't showing due to broken zoom level detection

MichaelReyes committed:
[23657] - implement clear non regional markers with debounce if 500ms
- add singleton class MapData and cache the regional stop marker options
- update MapViewModel to modify viewPort distinctUntilChanged and include zoom level checking so it'll still trigger even on the same region on a region level zoom
- local storage updates for schedule stops for Room db

MichaelReyes committed:
[23657] - zoom level stops fetcher logic update

MichaelReyes committed:
Merge branch 'refs/heads/develop' into task/23657-bottom-sheet-cards-refactoring

# Conflicts:
#	TripKitAndroidUI/src/main/java/com/skedgo/tripkit/ui/map/home/TripKitMapFragment.kt

MichaelReyes committed:
[23657] - additional merge conflicts fixes

MichaelReyes committed:
[23657] - additional merge conflicts fixes

MichaelReyes committed:
[23657] - additional merge conflicts fixes

MichaelReyes committed:
[23657] - search bar keyboarding visibility fix
- allow upward drag to extend bottom sheet when in half-expanded state
- wrap fragment layouts with proper scroll view to handle nested scrolling
- fix height allocation issues for child fragments in bottom sheet containers

MichaelReyes committed:
[23657] - showing/hiding markers based on viewport visible

MichaelReyes committed:
[23657] - only show StopPOILocation markers within viewport between zoom 12.1f–14.5f
- skip poiMarkers.showAll() in this zoom range to avoid blinking
- apply viewport filtering in the same frame for consistency
- code logs cleanup

MichaelReyes committed:
[24456] - zoom handling to show/hide city markers fix

MichaelReyes committed:
Merge branch 'refs/heads/develop' into task/23657-bottom-sheet-cards-refactoring

# Conflicts:
#	TripKitAndroidUI/src/main/res/values/strings.xml
#	tripkit-android

MichaelReyes committed:
[23657] merge conflicts fixes

MichaelReyes committed:
- update tripkit-android submodule reference
